File: retirodelrocioapp/lib/features/maintenance/domain/parts_request.dart
import 'package:flutter/foundation.dart';

/// A technician's request for a part against a work order — the Requests
/// tab's list (`GET /maintenance/parts-requests`).
@immutable
class PartsRequest {
  const PartsRequest({
    required this.id,
    required this.workOrderId,
    this.workOrderTitle,
    this.locationLabel,
    required this.partName,
    required this.quantity,
    this.note,
    required this.status,
    required this.statusLabel,
    this.requestedBy,
    this.createdLabel,
  });

  final int id;
  final int workOrderId;
  final String? workOrderTitle;
  final String? locationLabel;
  final String partName;
  final int quantity;
  final String? note;
  final String status;
  final String statusLabel;
  final String? requestedBy;
  final String? createdLabel;

  bool get isPending => status == 'pending';
  bool get isFulfilled => status == 'fulfilled';
  bool get isDenied => status == 'denied';

  factory PartsRequest.fromJson(Map<String, dynamic> json) => PartsRequest(
    id: (json['id'] as num?)?.toInt() ?? 0,
    workOrderId: (json['work_order_id'] as num?)?.toInt() ?? 0,
    workOrderTitle: json['work_order_title'] as String?,
    locationLabel: json['location_label'] as String?,
    partName: json['part_name'] as String? ?? 'Part',
    quantity: (json['quantity'] as num?)?.toInt() ?? 1,
    note: json['note'] as String?,
    status: json['status'] as String? ?? 'pending',
    statusLabel: json['status_label'] as String? ?? 'Pending',
    requestedBy: json['requested_by'] as String?,
    createdLabel: json['created_label'] as String?,
  );
}

File: retiro-del-rocio/app/Models/PartsRequest.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;

/** A technician's request for a part against a work order — the Requests tab's list. */
class PartsRequest extends Model
{
    public const PENDING = 'pending';

    public const FULFILLED = 'fulfilled';

    public const DENIED = 'denied';

    protected $attributes = [
        'status' => self::PENDING,
        'quantity' => 1,
    ];

    protected $fillable = [
        'work_order_id', 'part_name', 'quantity', 'note', 'status', 'requested_by', 'fulfilled_at',
    ];

    protected $casts = [
        'fulfilled_at' => 'datetime',
        'quantity' => 'integer',
    ];

    public function workOrder()
    {
        return $this->belongsTo(WorkOrder::class);
    }

    public function isPending(): bool
    {
        return $this->status === self::PENDING;
    }

    /** Pending → fulfilled. Idempotent. */
    public function fulfill(): bool
    {
        if (! $this->isPending()) {
            return false;
        }

        return $this->update(['status' => self::FULFILLED, 'fulfilled_at' => now()]);
    }

    /** Pending → denied. Idempotent. */
    public function deny(): bool
    {
        if (! $this->isPending()) {
            return false;
        }

        return $this->update(['status' => self::DENIED]);
    }

    public function statusLabel(): string
    {
        return match ($this->status) {
            self::FULFILLED => 'Fulfilled',
            self::DENIED => 'Denied',
            default => 'Pending',
        };
    }

    public function toMaintenanceArray(): array
    {
        $order = $this->relationLoaded('workOrder') ? $this->workOrder : $this->workOrder()->first();

        return [
            'id' => $this->id,
            'work_order_id' => $this->work_order_id,
            'work_order_title' => $order?->title,
            'location_label' => $order?->locationLabel(),
            'part_name' => $this->part_name,
            'quantity' => $this->quantity,
            'note' => $this->note,
            'status' => $this->status,
            'status_label' => $this->statusLabel(),
            'requested_by' => $this->requested_by,
            'created_label' => optional($this->created_at)->diffForHumans(),
        ];
    }
}
